dissect/target/plugin.py

```diff
@@ -264,607 +411,516 @@ def check_compatible(self) -> None:
         """
         raise NotImplementedError
 
-    def get_all_records(self) -> Iterator[Record]:
+    def __call__(self, *args, **kwargs) -> Iterator[Record | Any]:
         """Return the records of all exported methods.
 
         Raises:
             PluginError: If the subclass is not a namespace plugin.
         """
         if not self.__namespace__:
-            raise PluginError(f"Plugin {self.__class__.__name__} is not a namespace plugin")
+            raise PluginError(f"Plugin {self.__class__.__name__} is not a callable")
 
         for method_name in self.__exports__:
+            if method_name == "__call__":
+                continue
+
             method = getattr(self, method_name)
+            if getattr(method, "__output__", None) not in ("record", "yield"):
+                continue
 
             try:
                 yield from method()
             except Exception:
                 self.target.log.error("Error while executing `%s.%s`", self.__namespace__, method_name, exc_info=True)
 
-    def __call__(self, *args, **kwargs):
-        """A shortcut to :func:`get_all_records`.
 
-        Raises:
-            PluginError: If the subclass is not a namespace plugin.
-        """
-        if not self.__namespace__:
-            raise PluginError(f"Plugin {self.__class__.__name__} is not a callable")
-        return self.get_all_records()
+def register(plugincls: type[Plugin]) -> None:
+    """Register a plugin, and put related data inside :attr:`PLUGINS`.
 
+    This function uses the following private attributes that are set using decorators:
+        - ``__exported__``: Set in :func:`export`.
+        - ``__internal__``: Set in :func:`internal`.
 
-class OSPlugin(Plugin):
-    """Base class for OS plugins.
+    Additionally, ``register`` sets the following private attributes on the `plugincls`:
```

Comment:
(Architecture question, somewhat inarticulate)

Should OSPlugins also be plugins? 
Maybe they should be pluggable, but is this registry the right place to handle them?
Because it seems we need to make exceptions for them throughout the functions below.

Reply:
`find_plugin_functions` depends on `__functions__.__os__`.
And there is the `osfilter` in `plugins` and `functions` and `lookup`.

But I don´t have a solution, at this point I cannot even articulate the problem.
Only a nagging feeling.

Let's close for now.